### PR TITLE
[WIP] Test move add remove

### PR DESCRIPTION
**Pull request by ElectricMaxxx**
Merge when https://github.com/symfony-cmf/resource/pull/17 is merged.

Implemented tests on the editable repository created through the PR on resource component. reference change in composer.json will be removed.

### CODE REVIEW

**Comment by ElectricMaxxx** [comment]
Except the setUp() there is no funktional test for `PhpcrRepository`.

**Comment by ElectricMaxxx** [comment]
@dantleech any reasons why the PhpcrRepository has no functional tests in here?

**Comment by dantleech** [comment]
It is wrong, I was probably planning to move the tests to the `AbstractPhpcrTestCase` and extend both cases from that. The tests current tests would be identical for both. The setUps would be different.

**Comment by ElectricMaxxx** [comment]
Will try to do that. Still fixed the abstraction locally.

File: Tests/Functional/PhpcrRepositoryTest.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    /**
+     * @dataProvider provideMove
+     */
+    public function testRepositoryMove($sourcePath, $targetPath, $expectedNodeName)
+    {
+        $repository = $this->repositoryRegistry->get(self::REPOSITORY);
+        $res = $repository->move($sourcePath, $targetPath);
+
+        $this->assertEquals(1, $res);
+        $fooBar = $this->dm->find(null, '/test'.$targetPath);
```
Comment (ElectricMaxxx) [inline Tests/Functional/PhpcrRepositoryTest.php]: I thint the tool to validate if something is done does not matter, so i can look for the crated node by phpcr-odm too.

**Comment by ElectricMaxxx** [comment]
@Dantleech i currently do Not know how create a pure node instance on the easy way to pass it to the `PhpcResource` without touching the session.

File: Tests/Functional/CompositeRepositoryTest.php [hunk 2]
```diff
@@ -13,6 +13,8 @@
 
 class CompositeRepositoryTest extends PhpcrRepositoryTestCase
```
Comment (dantleech) [inline Tests/Functional/CompositeRepositoryTest.php]: Why extend the PhpcrTestCase here? CompositeRepository has nothing to do with PHPCR ...

File: Tests/Functional/PhpcrOdmRepositoryTest.php [hunk 3]
```diff
@@ -12,68 +12,98 @@
...
     public function testRepositoryFind($pattern, $nbResults)
     {
-        $repository = $this->repositoryRegistry->get('test_repository');
+        $repository = $this->repositoryRegistry->get(self::REPOSITORY);
         $res = $repository->find($pattern);
         $this->assertCount($nbResults, $res);
     }
+
+    /**
+     * @dataProvider provideMove
```
Comment (dantleech) [inline Tests/Functional/PhpcrOdmRepositoryTest.php]: What move cases do you need to test? I wouldn't add a `@dataProvider` before having a need for multiple scenarios, better to get a single test working first and then, if it makes sense, add a data provider.

File: Tests/Functional/PhpcrOdmRepositoryTest.php [hunk 4]
```diff
@@ -12,68 +12,98 @@
...
+     */
+    public function testRepositoryMove($sourcePath, $targetPath, $expectedNodeName)
```
Comment (dantleech) [inline Tests/Functional/PhpcrOdmRepositoryTest.php]: These tests could go into the AbstractTestCase. Could you use raw PHPCR to assert for both the PHPCR-ODM and PHPCR?

File: Tests/Functional/PhpcrOdmRepositoryTest.php [hunk 5]
```diff
@@ -12,68 +12,98 @@
...
+    {
+        $repository = $this->repositoryRegistry->get(self::REPOSITORY);
```
Comment (dantleech) [inline Tests/Functional/PhpcrOdmRepositoryTest.php]: I would just have a single, abstract, `getRepository` method in the abstract case which you would implement here:

```php
public function getRepository()
{
    return $this->repositoryRegistry->get('whatever');
}
```

Then there is no need for constants and the AbstractTestCase delegates the task of getting the Repository to the concrete test case.

File: Tests/Functional/PhpcrRepositoryTest.php [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        $this->markTestSkipped('Not able to test that atm.');
+    }
+
+    /**
+     * @dataProvider provideAdd
+     */
+    public function testRepositoryAddResourceCollection($path, $name)
+    {
+        $this->markTestSkipped('Not able to test that atm.');
```
Comment (dantleech) [inline Tests/Functional/PhpcrRepositoryTest.php]: You can only create PHPCR nodes through the PHPCR session with `$node = $session->getNode('foobar')->addNode('bar');` iirc.

**Comment by dantleech** [comment]
You can move much more to the AbstractPhpcrTestCase I think.

```php

class AbstractPhpcrTestCase extends \...
{
    private $session;

    public function setUp()
    {
        $this->session = // .. get PHPCR session
    }

    protected abstract function getRepository();

    public function testAddSomething()
    {
         $node = $this->session->getRoot()->addNode('foo');
         $resource = new PhpcrResource($node); // or whatever
         $this->getRepository()->add($resource, 'hello');

         $this->session->getNode('/hello');
    }

    public functoin testMoveSomething()
    {
         $this->getRepository()->move($resource, '/foo', '/bar');

         // PHPCR will throw an exception if /bar does not exist.
         $this->session->getNode('/bar');
    }

    // .. etc. All the tests are in this class.
}
```

File: Tests/Functional/PhpcrRepositoryTest.php (hunk 6 above)
Comment (ElectricMaxxx) [inline Tests/Functional/PhpcrRepositoryTest.php]: Yea, ok. But how to instantiate a PhpcrResource with a NodeInterface as second Parameter? When adding a PhpcrResource i can not create a node, fetch it and give it to the repository::add() to create the node again. Doing so the payload of that resource needs to get more freedom.

**Comment by ElectricMaxxx** [comment]
Sure. Mostly you see the equality when still written. I thing Get with the getNodeName()/getName() assert on the payload is the main difference.

File: Tests/Functional/PhpcrRepositoryTest.php (hunk 6 above)
Comment (dantleech) [inline Tests/Functional/PhpcrRepositoryTest.php]: Lets just skip the `add` for now and concentrate on the other methods. We can add `add` if necessary in a further PR.

File: Tests/Functional/CompositeRepositoryTest.php (hunk 2 above)
Comment (ElectricMaxxx) [inline Tests/Functional/CompositeRepositoryTest.php]: except from the setUp the abstract class is not related to PHPCR i just whanted to force to an equal test behavior as far as possible. I think as the repositories have same interfaces, so the test should have too.

File: Tests/Functional/PhpcrOdmRepositoryTest.php (hunk 4 above)
Comment (ElectricMaxxx) [inline Tests/Functional/PhpcrOdmRepositoryTest.php]: :+1:

File: Tests/Functional/PhpcrOdmRepositoryTest.php (hunk 5 above)
Comment (ElectricMaxxx) [inline Tests/Functional/PhpcrOdmRepositoryTest.php]: :+1:

File: Tests/Functional/PhpcrRepositoryTest.php (hunk 6 above)
Comment (ElectricMaxxx) [inline Tests/Functional/PhpcrRepositoryTest.php]: :+1: but we should keep that information in mind for the `PhpcrResource. I think we can't simply force to pass a `NodeInterface` in the constructor.

File: Tests/Functional/CompositeRepositoryTest.php (hunk 2 above)
Comment (dantleech) [inline Tests/Functional/CompositeRepositoryTest.php]: Does CompositeRepository implement EditableInterface? That would be out of the scope of this PR I think. In anycase it should not extend a PHPCR test case.

Comment (ElectricMaxxx) [inline Tests/Functional/CompositeRepositoryTest.php]: :+1: 

btw: that wasn't changed by me, but i will correct it.

File: Tests/Functional/CompositeRepositoryTest.php [hunk 7]
```diff
@@ -11,8 +11,10 @@
 
 namespace Symfony\Cmf\Bundle\ResourceBundle\Tests\Functional;
 
-class CompositeRepositoryTest extends PhpcrRepositoryTestCase
+class CompositeRepositoryTest extends RepositoryTestCase
```
Comment (ElectricMaxxx) [inline Tests/Functional/CompositeRepositoryTest.php]: I fixed the abstraction now. Both (or all three) repository tests do have one abstract method with the `setUp()`, cause even `CompositeRepositoryTest` needs some nodes/documents prepared:

```

        stuff:
            type: composite
            mounts:
                - { repository: test_repository_phpcr_odm, mountpoint: /content }
                - { repository: articles, mountpoint: /articles }
```
It works in that case `doctrine_phpcr_odm` type. So the composite repository shared the `setUp()` before my change and does it now. The name the extended name is cleare now and both phpcr repositories have its own abstraction to collect common methods now.

**Comment by ElectricMaxxx** [comment]
Can we do the major style forced changes (Licence date and []) in a separate PR and rebase mine on top?

File: Tests/Functional/RepositoryTestCase.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+ */
+abstract class RepositoryTestCase extends BaseTestCase
+{
+    /** @var DocumentManager */
+    protected $dm;
+
+    /** @var  RepositoryRegistryInterface */
+    protected $repositoryRegistry;
+
+    /** @var  Session */
```
Comment (dantleech) [inline Tests/Functional/RepositoryTestCase.php]:
```
/**
 * @var SessionInterface
 */

File: Tests/Functional/RepositoryTestCase.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * @author Maximilian Berghoff <Maximilian.Berghoff@mayflower.de>
+ */
+abstract class RepositoryTestCase extends BaseTestCase
+{
+    /** @var DocumentManager */
+    protected $dm;
+
+    /** @var  RepositoryRegistryInterface */
```
Comment (dantleech) [inline Tests/Functional/RepositoryTestCase.php]: Should be on separate lines, also you have two spaces:

```php
/**
 * @var RepositoryRegistryInterface
 */
```

File: Tests/Functional/PhpcrRepositoryTestCase.php [hunk 10]
```diff
@@ -11,32 +11,98 @@
...
+     */
+    public function testRepositoryGet($path, $expectedName)
+    {
+        $res = $this->getRepository()->get($path);
+        $this->assertNotNull($res);
+        $payload = $res->getPayload();
+
+        $this->assertEquals(
+            $expectedName,
+            ($payload instanceof Node ? $payload->getName() : $payload->getNodeName())
```
Comment (dantleech) [inline Tests/Functional/PhpcrRepositoryTestCase.php]: `s/Node/NodeInterface` - PHPCR is not Jackalope. PHPCR should always be type hinted with the PHPCR interfaces, i.e. `PHPCR\NodeInterface`.

**Comment by dantleech** [comment]
you can remove 5.3 and 5.4 support.

File: Tests/Functional/PhpcrOdmRepositoryTest.php [hunk 11]
```diff
@@ -12,68 +12,47 @@
...
-        $this->assertCount($nbResults, $res);
+        $document = new Generic();
+        $document->setNodeName($name);
+        $document->setParentDocument($this->dm->find(null, '/test'.('/' === $path ? '' : $path)));
+        $this->dm->persist($document);
+
+        $resource = new ArrayResourceCollection([new PhpcrOdmResource($path, $document)]);
+        $this->getRepository()->add($path, $resource);
+
+        $document = $this->dm->find(null, '/test'.$path.('/' === $path ? '' : '/').$name);
```
Comment (dantleech) [inline Tests/Functional/PhpcrOdmRepositoryTest.php]: concatenation should `. ' have ' . 'spaces`

Comment (dantleech) [inline Tests/Functional/PhpcrOdmRepositoryTest.php]: @dbu isn't there a PHPCR Util for the above?

**Comment by dantleech** [comment]
Applied StyleCI to master, you can rebase.

File: Tests/Functional/PhpcrOdmRepositoryTest.php (hunk 11 above)
Comment (dbu) [inline Tests/Functional/PhpcrOdmRepositoryTest.php]: actually symfony codestyle says to not have spaces in concatenation.

there is the PathHelper. you could do `'/test'.PathHelper::absolutizePath($name, $path)`

Comment (dantleech) [inline Tests/Functional/PhpcrOdmRepositoryTest.php]: ah well, thats probably why my PRs always need fixing on StyleCI :)

**Comment by ElectricMaxxx** [comment]
@dantleech 
I needed to squash for the rebas of 10 commits, but i got an esception for the phpcr configuration now:

``` 
Invalid configuration for repository "test_repository_phpcr"
 /home/maximilian/OpenSource/Cmf/resource-bundle/DependencyInjection/CmfResourceExtension.php:77
 /home/maximilian/OpenSource/Cmf/resource-bundle/DependencyInjection/CmfResourceExtension.php:50
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php:55
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/DependencyInjection/MergeExtensionConfigurationPass.php:39
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Compiler/Compiler.php:104
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/ContainerBuilder.php:593
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:511
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:133
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:146
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php:33
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony-cmf/testing/src/Functional/BaseTestCase.php:67
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony-cmf/testing/src/Functional/BaseTestCase.php:81
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony-cmf/testing/src/Functional/BaseTestCase.php:122
 /home/maximilian/OpenSource/Cmf/resource-bundle/vendor/symfony-cmf/testing/src/Functional/BaseTestCase.php:94
 /home/maximilian/OpenSource/Cmf/resource-bundle/Tests/Functional/RepositoryTestCase.php:41
```

The option resolver throws:
```
The option "basepath" does not exist. Defined options are: "basedir".
``
had there been a change on the configuration?

**Comment by dantleech** [comment]
It should be `basepath`. Looks like I made a bug, will try and fix it
this afternoon.

**Comment by dantleech** [comment]
Fixed, you can rebase again ...

**Comment by ElectricMaxxx** [comment]
Oh. That was really fast. Will do it tonight.

**Comment by dantleech** [comment]
Replaced by #31

**Pull request closed without merging**